presto-docs/src/main/sphinx/admin/verifier.rst

```diff
@@ -286,6 +288,9 @@ Name                                        Description
 ``running-mode``                            Set to ``query-bank`` to make the Verifier run in ``query-bank`` mode. Supports
                                             ``query-bank`` and ``control-test``. Defaults to ``control-test``.
 ``save-snapshot``                           Set to ``true`` to save checksums to ``mysql`` database.
+``nondeterministic-function-substitutes``   Specification of substituting nondeterministic functions with deterministic functions,
+                                            in the format of ``/func_x(c_0,c_1)/func_y(c_1)/`` where ``func_x`` would be
```

Comment:
Wondering if it would be mor readable as 

```
/foo(_,y)/bar(y)/
```

How would one specify multiple replacements?

Let's also provide some examples, approx_distinct -> cnt; approx_percentile -> avg.

Reply:
Can you test what happens is function replacements are not valid, e.g. replacement function doesn't support input types or returns different result type? Let's make sure errors are clear in these cases.

Let's also explain how one can tell which query was verified as is vs. after re-write in the results.

Reply:
Masha, I got it that we need to figure out a way to verify the function signature. I think right now if the replacement func is not legal, the individual test cases with the replaced functions will fail the standard syntax check.

I think the best way to verify func signature is to use the class FunctionAndTypeManager. But the difficulty here is here in verifier we don't have the context to create a FunctionAndTypeManager object as we have in the query execution flow.
https://github.com/prestodb/presto/blob/a7e8ca39020c76a0c329a0197d243353a0a01502/presto-main/src/main/java/com/facebook/presto/metadata/FunctionAndTypeManager.java#L131-L137

Have been looking at it. Need some more time.

Reply:
> Can you test what happens is function replacements are not valid, e.g. replacement function doesn't support input types or returns different result type? Let's make sure errors are clear in these cases.
> 
> Let's also explain how one can tell which query was verified as is vs. after re-write in the results.

> If we have access to return types then we can test them, otherwise we can just let query to fail. If such failures happen rarely it is still better that way, since otherwise we would skip them anyway.
> 
> As for detecting an "as is" query vs "altered" one - can we add some comment to the end of the query with the replacements which went through? Like `approx_distinct(city, 0.05) -> count(city) `


So for illegal func replacements, what I add so far is

1) Let it run, and the CONTROL query will fail, as the illegal replacement is also made to the CONTROL query. When the CONTROL query fails, the Verifier result is supposed to SKIPPED as today. I will set the "SkippedReason" field to "POSSIBLE_INVALID_FUNCTION_SUBSTITUTE". It will show on the UI.

2) For surfacing "as is" query vs "altered" one, I appended a comment to the end of query text that says what replacements have taken place, as below ex.. It will show up on the UI.
> -- "map_agg"(app_id, user_age) is substituted with "max_by"(user_age)